service/src/BatchServer.cpp

```diff
@@ -0,0 +1,258 @@
+        }
+        if (!is_test) {
+            auto setup = [this]() {this->create_shmem();};
+            auto run = [this]() {this->run_batch();};
+            m_server_pid = fork_with_setup(setup, run);
+        }
+        m_is_active = true;
+    }
+
+    BatchServerImp::~BatchServerImp()
+    {
+        stop_batch();
+    }
+
+    int BatchServerImp::server_pid(void) const
+    {
+        return m_server_pid;
+    }
+
+    std::string BatchServerImp::server_key(void) const
+    {
+        return m_server_key;
+    }
+
+    void BatchServerImp::stop_batch(void)
+    {
+        if (m_is_active) {
+            m_batch_status->send_message(M_MESSAGE_QUIT);
+            m_batch_status->receive_message(M_MESSAGE_CONTINUE);
+            m_is_active = false;
+        }
+    }
+
+    void BatchServerImp::run_batch(void)
+    {
+        push_requests();
+        m_batch_status->send_message(M_MESSAGE_CONTINUE);
+        // Start event loop
+        bool do_cont = true;
+        while (do_cont) {
+            int message = m_batch_status->receive_message();
+            switch (message) {
+                case M_MESSAGE_READ:
+                    read_and_update();
+                    break;
+                case M_MESSAGE_WRITE:
+                    update_and_write();
+                    break;
+                case M_MESSAGE_QUIT:
+                    do_cont = false;
+                    break;
+                default:
+                    throw Exception("BatchServerImp::run_batch(): Received unknown response from client: " +
+                                    std::to_string(message), GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+                    break;
+            }
+            m_batch_status->send_message(M_MESSAGE_CONTINUE);
+        }
+    }
+
+    bool BatchServerImp::is_active(void) const
+    {
+        return m_is_active;
+    }
+
+    void BatchServerImp::push_requests(void)
+    {
+        for (const auto &req : m_signal_config) {
+            m_signal_idx.push_back(
+                m_pio.push_signal(req.name, req.domain, req.domain_idx));
+        }
+        for (const auto &req : m_control_config) {
+            m_control_idx.push_back(
+                m_pio.push_control(req.name, req.domain, req.domain_idx));
+        }
+    }
+
+    void BatchServerImp::read_and_update(void)
+    {
+        if (m_signal_config.size() == 0) {
+            return;
+        }
+
+        m_pio.read_batch();
+        double *buffer = (double *)m_signal_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &idx : m_signal_idx) {
+            buffer[buffer_idx] = m_pio.sample(idx);
+            ++buffer_idx;
+        }
+    }
+
+    void BatchServerImp::update_and_write(void)
+    {
+        if (m_control_config.size() == 0) {
+            return;
+        }
+
+        double *buffer = (double *)m_control_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &idx : m_control_idx) {
```

Comment:
I would like to rename `idx` to `handle`. And rename `m_control_idx` to `m_control_handle`.